### DISCUSSION TITLE
How well / fully featured (roughly) does the Intel ARC 770 work vs. a comparable NVIDIA GPU e.g. 3060 12G?

### DISCUSSION THREAD

**Discussion by ghchris2021** [vladmandic/sdnext#2024] [Q&A] [answered]
Firstly congratulations to the developers for making this excellent free software and also working on making it cross-platform for the OS and GPU!

I am wondering qualitatively how well / fully featured (roughly) does the Intel ARC 770 work vs. a comparable NVIDIA GPU e.g. 3060 12G?

The main reason I ask is I was glad to see both ARC and NVIDIA LINUX support but haven't tried it and am deciding what to put effort into setting up / trying and what roughly to expect.  Making my curiosity about possible limitations with the ARC higher is that I ran into another thread (initial fact finding comparing how well SD/SDXL and various SD/XL UIs support ARC now and what might work best at this time) which suggested that there were some severe problems that had been recently posited / encountered by users trying to work with ARC A7xx using ComfyUI and they suggested citations from various Intel related forums that the root cause was in Intel libraries and would be a significant limitation until the VRAM usage / allocation / addressing issue(s) are fixed upstream.
e.g.
https://github.com/comfyanonymous/ComfyUI/discussions/476
"...3.) Intel Arc right now has a huge problem with allocating more than 4GB of VRAM in IPEX even though the card has more VRAM in the case of the A750/A770..."

https://github.com/intel/intel-extension-for-pytorch/issues/325
https://github.com/intel/compute-runtime/issues/627
https://github.com/oneapi-src/oneDNN/issues/1638

So I gather that across various APIs there may be some "32-bit" addressing limit and various other allocation limits and such that make even the A770/16GB not work with SD / NN / CL sorts of work loads if processing any given data structure or maybe collection of simultaneously adressable data structures (?) >= 4GBy, and that if there is a work-around it may have to do with limiting the data block / model / operand sizes to less than that which I gather it isn't needed for NVIDIA.

But then I saw this here:
https://github.com/vladmandic/automatic/discussions/1588#discussioncomment-6415290
"...I successfully generated 100 pictures without running out of VRAM or system memory...."

Which though it suggests a necessary work-around for a bad VRAM leak seemed hopeful in that I didn't see another
implication of running out of or having insufficient available VRAM for still other reasons.

So long story short I was just wondering how well it works in short from a feature / function / stability / capacity standpoint from an end user's perspective vs. say the comparable NV 3060/12G or some such thing?  

Do those 4GBy chunk limits cause major limitations or slow-downs in this setup or needed significant work-arounds / ?

**Comment by vladmandic** [comment, marked as answer]
> Do those 4GBy chunk limits cause major limitations or slow-downs in this setup or needed significant work-arounds / ?

Not in normal workflows, some in SD-XL workflows, but mitigated with VAE slicing/tiling without too much of performance loss.
For raw numbers, best to take a look at system info benchmark database: https://vladmandic.github.io/sd-extension-system-info/pages/benchmark.html

**Reply by ghchris2021** [reply]
Thank you very much, that's very useful and encouraging information.  I'll give it a try and see how well I can make it work!
Thanks again for creating this project & SW!

**Comment by Disty0** [comment]
VRAM leak problem was a WSL only issue. It doesn't exist in Native Linux.
And WSL VRAM leak issue got worked around by disabling torch.xpu.empty_cache if WSL is detected.
torch.xpu.empty_cache was leaking VRAM instead of clearing it on WSL.
It won't clear the unused VRAM in WSL anymore but VRAM usage won't grow.
And PyTorch can reuse those unused VRAM space without allocating more.


4GB problem is a different issue and got worked around by these commits: https://github.com/vladmandic/automatic/commit/9d17cf4c122b98b25b8cb9e3388c1a75df68cdb2 & https://github.com/vladmandic/automatic/commit/489d0382cfd95ba4420d94dce4d2432c0cb54dd8
Dynamically slicing the attention layers to keep it under 4GB works without any noticeable performance loss.
And as a side effect of this, A770 16 GB can generate 4096x4096 SDXL or SD 1.5 images with Attention Slicing, VAE Tiling and Model Shuffling turned on. 
You will get this 4GB error if you try to generate above 5120x5120 (you will run out of VRAM even if you could).

Performance wise; 
A770 is between a RTX 3060 and a RTX 3060 Ti on low resolutions.
A770 is equivalent to a RTX 3070 on higher resolutions or batch sizes.
ARC GPUs doesn't like lower resolutions.


And just for fun, a direct 4096x4096 SDXL image generated on an A770 16GB:
![01818-astreapixieXLAnime_v16-4170966426](https://github.com/vladmandic/automatic/assets/47277141/e95305d2-d101-4cc8-a916-d0f5b7c0c593)

And this is with 1920x1080 as base and re-generated at 3840x2160 with Img2Img:
![00062-astreapixieXLAnime_v16-4194632960](https://github.com/vladmandic/automatic/assets/47277141/4157339e-becb-4030-abc4-0511a6c2ec31)

**Reply by cyatarow** [reply]
> with Attention Slicing, VAE Tiling and Model Shuffling turned on

Where is "Model Shuffling" located in the settings?

**Reply by Disty0** [reply]
These 3 options:
https://github.com/vladmandic/automatic/wiki/SD-XL#option-1-model-shuffle
![image](https://github.com/vladmandic/automatic/assets/47277141/62070ce2-eb32-43e3-a142-3db2db2e046c)


Settings i use:
![image](https://github.com/vladmandic/automatic/assets/47277141/5b1e823a-3009-4cac-8712-9c4037229176)

**Reply by cyatarow** [reply]
Thank you!
By the way, these are located in "Diffusers Settings" section.
Do they take effect when I don't use the Diffusers backend?

**Reply by Disty0** [reply]
They are for Diffusers only. 
You can generate up to 2032x2032 using original backend without messing with sub-quad chunk sizes.
Even if you manage to go above 2032x2032 with original backend, it will be extremely slow anyway.
Original backend breaks apart at high resolutions and high demanding use cases.
Use Diffusers for high resolutions (or anything in general).

**Reply by cyatarow** [reply]
I see.
Unfortunately, I've used a model not distributed in Diffusers format, so I'm not going to be able to benefit from these settings.

**Reply by Disty0** [reply]
What model is that?
A single safetensors file works perfectly fine on Diffusers.
And SD 1.5 works better on Diffusers too.

**Reply by cyatarow** [reply]
Sorry, I misunderstood.
Models in safetensors format work fine.

(And actually, I've used SD 1.5 models, not SDXL ones. But I was curious, so I asked.)

**Reply by Disty0** [reply]
Training TI and HyperNetworks in SD.Next is working now.

Training on Kohya SS also works with my fork:

https://github.com/Disty0/kohya_ss/tree/master

https://github.com/bmaltais/kohya_ss/pull/1499
